agent/lib/kontena/rpc_client.rb

```diff
@@ -34,26 +37,53 @@ def notification(method, params)
       logger.error exc.message
     end
 
+    # This method should not raise, or the Actor will crash, and terminate any other pending requests.
+    #
     # @param [String] method
     # @param [Array] params
-    # @return [Object]
-    def request(method, params)
+    # @param [Fixnum] timeout seconds
+    # @return [Object, Exception]
+    def request_with_error(method, params, timeout: 30)
       id = request_id
       @requests[id] = nil
-      sleep 0.01 until @client.connected?
+
+      if !wait_until("websocket client is connected", timeout: timeout, interval: 0.01) { @client.connected? }
+        return nil, TimeoutError.new(500, 'WebsocketClient is not connected')
+      end
+
       @client.send_request(id, method, params)
-      time = Time.now.utc
-      until !@requests[id].nil?
-        sleep 0.01
-        raise TimeoutError.new(500, 'Request timed out') if time < (Time.now.utc - 30)
+
+      if !wait_until("request #{id} has response", timeout: timeout, interval: 0.01) { @requests[id] }
+        return nil, TimeoutError.new(500, 'Request timed out')
       end
+
       result, error = @requests.delete(id)
+
+      if error
+        return result, Error.new(error['code'], error['message'])
+      else
+        return result, nil
+      end
+    end
+
+    # Async request wrapper.
+    #
+    # Logs a warning and returns nil on errors.
+    # Use Kontena::Helpers::RpcError.rpc_request to get a raised error instead.
+    #
+    # @return [Object, nil]
+    def request(method, params, **opts)
+      result, error = request_with_error(method, params, **opts)
+
       if error
-        raise Error.new(error['code'], error['message'])
+        warn "RPC request #{method} failed: #{error}"
```

Comment:
This is a bit awkward, but it's still an improvement over how `rpc_client.async.request(...)` behaved before... it would not log any errors.